### ISSUE TITLE
ES Module usage from Node?

### ISSUE THREAD

**Issue by deklanw** [graphology/graphology#382]
index.js:

```
import {UsageGraphError} from "graphology";
```

Running `node index.js` I get

```
import {UsageGraphError} from "graphology";
        ^^^^^^^^^^^^^^^
SyntaxError: Named export 'UsageGraphError' not found. The requested module 'graphology' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'graphology';
const {UsageGraphError} = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:127:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:191:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:337:24)
    at async loadESM (node:internal/process/esm_loader:88:5)
    at async handleMainPromise (node:internal/modules/run_main:61:12)

Node.js v17.4.0
```

Here's the package.json, note the `"type": "module",`
```
{
  "name": "module_testing",
  "type": "module",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "graphology": "^0.24.1"
  }
}
```

**Comment by deklanw** [comment]
Looks like you need "Conditional Exports" https://nodejs.org/api/packages.html#conditional-exports

Tried adding this to graphology's package.json

```
  "type": "module",
  "exports": {
    "require": "./dist/graphology.cjs.js",
    "default": "./dist/graphology.esm.js"
  },
```

And everything works if I pass the right flag to node:
`node --experimental-specifier-resolution=node index.js`

The flag seems to be necessary because of eliding `.js` extensions, like:

```javascript
import Iterator from 'obliterator/iterator';
import take from 'obliterator/take';
import chain from 'obliterator/chain';
```

**Comment by Yomguithereal** [comment]
It's the `type: module` than frightens me as I suspect it might break everyone else indeed. Do you think exporting a standalone `.mjs` file in the dist folder (while including deps, which is a bit silly) could work? JS packaging has become such a mess (ok maybe it always was). I am unsure how I can keep supporting every system. Let me document myself a little bit.

**Comment by schickling** [comment]
I was facing the same problem. It seems like `graphology` basically can't be use with TypeScript and modern Node.js. I've worked around the problem using the following local patch file but hope `graphology` and its dependencies will adopt modern ESM practices soon 🙏 

```patch
diff --git a/dist/graphology.esm.js b/dist/graphology.esm.js
index aa67724e820dc5c46106feb6c6bc8c2df898eb35..a0e34226947cc47c99aa3d59dcb5ed133a86c3d5 100644
--- a/dist/graphology.esm.js
+++ b/dist/graphology.esm.js
@@ -1,7 +1,11 @@
 import { EventEmitter } from 'events';
-import Iterator from 'obliterator/iterator';
-import take from 'obliterator/take';
-import chain from 'obliterator/chain';
+import obliterator from 'obliterator'
+
+// import Iterator from 'obliterator/iterator';
+// import take from 'obliterator/take';
+// import chain from 'obliterator/chain';
+
+const { Iterator, take, chain } = obliterator
 
 /**
  * Graphology Utilities
diff --git a/package.json b/package.json
index 234d6a75ad64d0f36f2952eaa7c4428a01f3f248..aca10d8625384bb95b369da1eba38be10a50a574 100644
--- a/package.json
+++ b/package.json
@@ -1,11 +1,13 @@
 {
   "name": "graphology",
   "version": "0.25.1",
+  "type": "module",
   "description": "A robust and multipurpose Graph object for JavaScript.",
-  "main": "dist/graphology.cjs.js",
-  "module": "dist/graphology.esm.js",
-  "browser": "dist/graphology.umd.min.js",
   "types": "dist/graphology.d.ts",
+  "exports": {
+    "require": "./dist/graphology.cjs.js",
+    "default": "./dist/graphology.esm.js"
+  },
   "scripts": {
     "clean": "rimraf dist specs",
     "build": "npm run clean && rollup -c && babel tests --out-dir specs && cp src/endpoint.esm.d.ts dist/graphology.d.ts",

```

**Comment by Yomguithereal** [comment]
@schickling is the JS module schism over yet? What would your solution entail regarding backwards compatibility? What would be the minimal node/npm version supported then?

**Comment by schickling** [comment]
I'd say by now new/well-maintained packages are ESM-first. Targeting Node 16+ feels reasonable to me.

**Comment by Yomguithereal** [comment]
That does not feel reasonable to me. I know for a fact that `graphology` and its standard library are relied upon by dependencies and platforms that cannot upgrade beyond Node 14 or web packagers that don't support the latest ESM-first scheme (not ESM itself, which is what I find baffling regarding what I call some kind of "schism" here). I wouldn't call those legacy as some of them are not even two years old. My goal has always been to make sure everyone can consume my libraries, but here it seems I must make a choice and leave some people behind (some people that often upgrade the graphology related packages for perf reasons mostly, but not their platforms, for reasons beyond my control and this is also personally my case). What's more, I have two more problems: 1. I maintain hundreds of Open Source libraries and I don't have time to make significant changes to their codebases or packaging and 2. and I usually want to avoid transpilation (concatenation is fine) because it tends to introduce subtle issues wrt JIT optimizations and I want to retain control over performance as much as possible.

`</rant>` Now, I understand your concern and I would like to upgrade but I cannot find a way to do so without breaking things for people that upgrade the package regularly (including myself). And since I am sure you know much more than I do wrt the recent changes in package.json/packaging and ESM-first stuff, I am sure we can find something? How is the `"type": "module"` thing understood by older versions of node? Would it break? Would it be ignored? Can't we keep a module type + `exports` + `main` `module` etc. ? Can we play with the `.mjs` extensions somehow?

Basically is there some kind of accepted answer nowadays that let people keep backwards compability all while adapting to ESM-first in a node package? Last time I checked it was impossible (which feels completely bonkers to me) and this is why I did not do anything. Maybe the situation has evolved?

**Comment by Yomguithereal** [comment]
After reading a little bit about this: would it work not to set `type: module` but add an `exports` to the package all while providing a `.mjs` version of the lib along with the older cjs and esm ones?

**Comment by schickling** [comment]
I appreciate and share your perspective above. It's unfortunate - at the same time I'm glad things are being "improved" assuming you're using a more modern foundation.


> After reading a little bit about this: would it work not to set `type: module` but add an `exports` to the package all while providing a `.mjs` version of the lib along with the older cjs and esm ones?

Yes, I think that should work! Note (i.e. see my `.diff`) that this also applies to the other peer and dependency packages.

**Comment by Yomguithereal** [comment]
@schickling if I roll a tentative alpha version from a branch with our ideas will you be able to test it on older and newer nodes to make sure it works both in ESM and require?

> P.S. I've also pinged you on Discord. Happy to continue chatting there e.g. to try out a preview release.

I am not sure I still have an account on the platform.

**Comment by schickling** [comment]
> @schickling if I roll a tentative alpha version from a branch with our ideas will you be able to test it on older and newer nodes to make sure it works both in ESM and require?

I'll be able to help test it for newer Node versions. 👍

**Comment by Yomguithereal** [comment]
@schickling can you try the `0.26.0-alpha1` I just published on npm?

**Comment by schickling** [comment]
Seems to work for me 🙌

**Comment by schickling** [comment]
Speaking of: I've just seen this error related to `obliterator`: 

<img width="1107" alt="CleanShot 2023-03-02 at 18 21 25@2x" src="https://user-images.githubusercontent.com/1567498/222504392-3a0e9b72-410c-469f-9cea-92863b3827e1.png">

**Comment by Yomguithereal** [comment]
Wait is this shit viral :) ? Must all the dependencies used by graphology support ESM-first? Your diff is alright functionally but is not equivalent wrt file size because you are now importing **all** `obliterator` and not just the necessary parts. I believe in tree-shaking as well as the next guy but it seems to often fail miserably. Let me see what I can do.

**Comment by schickling** [comment]
Any update on this @Yomguithereal? The preview release + my patch seem to work fine :)

**Comment by Yomguithereal** [comment]
@schickling I don't remember what is blocking here? `obliterator` must be ESM-first compatible? If so I will need some time because it is downloaded 1.5M per week on npm and I cannot afford to break anything.

**Comment by schickling** [comment]
> `obliterator` must be ESM-first compatible?

Yes. I understand. Thanks for your effort! 🙏

**Comment by Ritik-Banger-Biz4group** [comment]
Any update on this? 
`graphology-6f8bbb11.js:1 Uncaught TypeError: Cannot read properties of undefined (reading 'EventEmitter')`
It is crashing the whole react app. 

![image](https://github.com/graphology/graphology/assets/104008405/1f52fb15-5310-40ee-9e09-fa842484589e)

**Comment by kaisen-ai** [comment]
What is the progress for this issue? This issue is affecting our deployment CI/CD.

**Comment by Yomguithereal** [comment]
@kaisen-ai as explained in #489 newest v0.25.4 reverted to the original package.json so this should not affect your CI/CD anymore except if you are hellbent on consuming this package as ESM-first which it never was. 

Regarding ESM-first integration I am somewhat stuck because graphology relies on critical dependencies which are not yet cjs/esm hybrid and they risk breaking much more CI/CD than yours if not done correctly :)

**Comment by dmurvihill** [comment]
Two years on and most Node packages have completed the transition to ESM, so this issue may finally be ripe. I have a [draft fix](https://github.com/Yomguithereal/obliterator/pull/35/files) open in the Obliterator repo for the import issue there. Would appreciate testing by interested parties.

**Comment by Yomguithereal** [comment]
Someone also opened a PR #529 removing `obliterator` from the core, which can make all this more straightforward (still need obliterator for the standard library though). I am waiting for the requested modifications on the PR to enact the change.
